src/Futhark/IR/Mem/Interval.hs

```diff
@@ -0,0 +1,137 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Futhark.IR.Mem.Interval
+  ( Interval (..),
+    distributeOffset,
+    expandOffset,
+    intervalOverlap,
+    selfOverlap,
+    primBool,
+    intervalPairs,
+    justLeafExp,
+  )
+where
+
+import Data.Function (on)
+import Data.List (maximumBy, minimumBy, (\\))
+import Futhark.Analysis.AlgSimplify qualified as AlgSimplify
+import Futhark.Analysis.PrimExp.Convert
+import Futhark.IR.Prop
+import Futhark.IR.Syntax hiding (Result)
+import Futhark.Util
+
+data Interval = Interval
+  { lowerBound :: TPrimExp Int64 VName,
+    numElements :: TPrimExp Int64 VName,
+    stride :: TPrimExp Int64 VName
+  }
+  deriving (Show, Eq)
+
+instance FreeIn Interval where
+  freeIn' (Interval lb ne st) = freeIn' lb <> freeIn' ne <> freeIn' st
+
+distributeOffset :: MonadFail m => AlgSimplify.SofP -> [Interval] -> m [Interval]
+distributeOffset [] interval = pure interval
+distributeOffset offset [] = fail $ "Cannot distribute offset " <> show offset <> " across empty interval"
+distributeOffset offset [Interval lb ne 1] = pure [Interval (lb + TPrimExp (AlgSimplify.sumToExp offset)) ne 1]
+distributeOffset offset (Interval lb ne st0 : is)
+  | st <- AlgSimplify.Prod False [untyped st0],
+    Just (before, quotient, after) <- focusMaybe (`AlgSimplify.maybeDivide` st) offset =
+      distributeOffset (before <> after) $
+        Interval (lb + TPrimExp (AlgSimplify.sumToExp [quotient])) ne st0 : is
+  | [st] <- AlgSimplify.simplify0 $ untyped st0,
+    Just (before, quotient, after) <- focusMaybe (`AlgSimplify.maybeDivide` st) offset =
+      distributeOffset (before <> after) $
+        Interval (lb + TPrimExp (AlgSimplify.sumToExp [quotient])) ne st0 : is
+  | otherwise = do
+      rest <- distributeOffset offset is
+      pure $ Interval lb ne st0 : rest
+
+findMostComplexTerm :: AlgSimplify.SofP -> (AlgSimplify.Prod, AlgSimplify.SofP)
+findMostComplexTerm prods =
+  let max_prod = maximumBy (compare `on` (length . AlgSimplify.atoms)) prods
+   in (max_prod, prods \\ [max_prod])
+
+findClosestStride :: [PrimExp VName] -> [Interval] -> (PrimExp VName, [PrimExp VName])
+findClosestStride offset_term is =
+  let strides = map (untyped . stride) is
+      p =
+        minimumBy
+          ( compare
+              `on` ( (\(AlgSimplify.Prod _ xs) -> length (offset_term \\ xs))
```

Comment:
Move this big lambda into a named helper function.

Reply:
Done